src/dialogxml/dialogs/dialog.cpp

```diff
@@ -508,16 +508,31 @@ void cDialog::run(std::function<void(cDialog&)> onopen){
 			currentFocus = iter->first;
 		}
 	}
-	// Sometimes it seems like the Cocoa menu handling clobbers the active rendering context.
-	// For whatever reason, delaying 100 milliseconds appears to fix this.
-	sf::sleep(sf::milliseconds(100));
-	// So this little section of code is a real-life valley of dying things.
-	// Instantiating a window and then closing it seems to fix the update error, because magic.
-	win.create(sf::VideoMode(1,1),"");
-	win.close();
-	win.create(sf::VideoMode(winRect.width(), winRect.height()), "Dialog", sf::Style::Titlebar);
-	winLastX = parentPos.x + (int(parentSz.x) - winRect.width()) / 2;
-	winLastY = parentPos.y + (int(parentSz.y) - winRect.height()) / 2;
+	if (ui_scale<=0) {
+		ui_scale = get_float_pref("UIScale", 1.0);
+		if (ui_scale < 0.1) ui_scale = 1.0;
+	}
+
+	// check that the dialog can be displayed on the screen
+	sf::VideoMode desktop = sf::VideoMode::getDesktopMode();
+	if (desktop.width< ui_scale*winRect.width())
+		ui_scale=float(desktop.width)/winRect.width();
+	// we need keep enough space for the mac menu and the dialog title
+	// and add more space on a retina mac
+	// checkme: we need probably to also add getMenuHeight at least on Linux
```

Comment:
I'd like to accept this change, but I have three requests:

* Unless you're certain it _creates_ an issue, I'd like to keep the hack that sleeps for 100ms and creates and destroys a 1x1 window. If it's not causing any problems and _might_ be fixing an issue for someone, it shouldn't just be deleted.
* There's no need to call getMenuHeight because only the main window has menus. (Edit menu shortcuts like copy and paste work independently of the actual menu.) Therefore, you can delete the checkme comment.
* In e47b7e6a223a4fbc666e04138f9b93aefe2b1002 I allowed differing the UI scale in the game and in the editors, so that will need to be handled somehow. Passing UI scale as a parameter in the constructor is one possibility, though that carries the annoying task of updating every single dialog construction in the entire source to pass it. Another possibility would be a `get_ui_scale()` function that's implemented differently in each program.

I'd also like to better understand what you're doing here to limit the size to something that fits on the screen. You're just scaling it by less than the expected 2x in that case, aren't you? I wonder if this is a good approach… doesn't 1.5x look a lot worse than 2x?

Reply:
> I'd like to accept this change, but I have ~three~two requests:
> 
> * Unless you're certain it _creates_ an issue, I'd like to keep the hack that sleeps for 100ms and creates and destroys a 1x1 window. If it's not causing any problems and _might_ be fixing an issue for someone, it shouldn't just be deleted.

I know that before I changed this code completely, the hack was necessary on my mac and with the new code, it is not necessary anymore. So I guess I solved this problem but we can keep the hack in a dead code (with #if 0 #endif), so we can reactivate it if necessary.

> * There's no need to call getMenuHeight because only the main window has menus. (Edit menu shortcuts like copy and paste work independently of the actual menu.) Therefore, you can delete the checkme comment.

Ok

> * ~In [e47b7e6](https://github.com/calref/cboe/commit/e47b7e6a223a4fbc666e04138f9b93aefe2b1002) I allowed differing the UI scale in the game and in the editors, so that will need to be handled somehow. Passing UI scale as a parameter in the constructor is one possibility, though that carries the annoying task of updating every single dialog construction in the entire source to pass it. Another possibility would be a `get_ui_scale()` function that's implemented differently in each program.~
>
 
I have to investigate. Actually, in my code, it's also possible, at least on Mac, ie. each executable uses uses a different preference file. We'd have to check if this is true on Windows and on Linux.

> I'd also like to better understand what you're doing here to limit the size to something that fits on the screen. You're just scaling it by less than the expected 2x in that case, aren't you? I wonder if this is a good approach… doesn't 1.5x look a lot worse than 2x?

Yes, the appearance is a bit worse, but having to reduce the size of a dialog box almost never happens. So in practice, using a size closed to the desired size seems ok....

Reply:
> I have to investigate. Actually, in my code, it's also possible, at least on Mac, ie. each executable uses uses a different preference file. We'd have to check if this is true on Windows and on Linux.

Ah, yes, we'd have to make sure they also use a separate preference file on Windows and Linux. I reverted it because I noticed they already use a separate preference file on the Mac, but I'd have to check if that's also true on Windows and Linux.

> Yes, the appearance is a bit worse, but having to reduce the size of a dialog box almost never happens. So in practice, using a size closed to the desired size seems ok....

Have you seen cases where the dialog's size needs to be reduced when set to 2x? (I won't worry about when it's set to 4x.)

Reply:
I thought "View Changes" on this conversation took you to the commit I'd commented on, but apparently that's not true, so here it is: 5286e798408c66433949a5a5bf26c462e4a45403

Reply:
On my macbook air (and on an iMac), I use 4x scaling. I think BoE preferences are scaled down a bit on my MacBook Air (maybe by a factor of 20%) but other than that, I rarely see scaled dialogs. So with 2x scaling, I'd say it should never happen.

Reply:
So pick spell and training dialogs are not scaled down even at 4x?

Reply:
I'll add a breakpoint tomorrow to give an accurate answer. But if I remember what happened before I added this code to scale down the overly large dialogs, the spellcasting dialog was a bit too big (i.e., the ok button didn't show up completely), so it's probably also scaled down a bit. I don't remember if there was problem with the training dialogue.

Reply:
So I add a message when the dialogs are reduced: the preferences dialog is reduced ( uiScale=2.98), the casting and training dialogs are not reduced.
<img width="625" alt="Capture d’écran 2023-01-24 à 18 25 29" src="https://user-images.githubusercontent.com/13163168/214517473-f91e573c-db25-4f91-9420-20ed3a044d37.png">